sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java

```diff
@@ -433,4 +589,178 @@ public void updateWatermark(Instant watermark) {
+      throw new UnsupportedOperationException(
+          "Cannot access element outside of @ProcessElement method.");
+    }
+
+    @Override
+    public TimeDomain timeDomain(DoFn<InputT, OutputT> doFn) {
+      return timeDomain();
+    }
+
+    @Override
+    public OutputReceiver<OutputT> outputReceiver(DoFn<InputT, OutputT> doFn) {
+      return DoFnOutputReceivers.windowedReceiver(this, null);
+    }
+
+    @Override
+    public OutputReceiver<Row> outputRowReceiver(DoFn<InputT, OutputT> doFn) {
+      return DoFnOutputReceivers.rowReceiver(this, null, mainOutputSchemaCoder);
+    }
+
+    @Override
+    public MultiOutputReceiver taggedOutputReceiver(DoFn<InputT, OutputT> doFn) {
+      return DoFnOutputReceivers.windowedMultiReceiver(this);
+    }
+
+    @Override
+    public DoFn<InputT, OutputT>.OnTimerContext onTimerContext(DoFn<InputT, OutputT> doFn) {
+      return this;
+    }
+
+    @Override
+    public RestrictionTracker<?, ?> restrictionTracker() {
+      throw new UnsupportedOperationException("RestrictionTracker parameters are not supported.");
+    }
+
+    @Override
+    public State state(String stateId) {
+      StateDeclaration stateDeclaration = context.doFnSignature.stateDeclarations().get(stateId);
+      checkNotNull(stateDeclaration, "No state declaration found for %s", stateId);
+      StateSpec<?> spec;
+      try {
+        spec = (StateSpec<?>) stateDeclaration.field().get(context.doFn);
+      } catch (IllegalAccessException e) {
+        throw new RuntimeException(e);
+      }
+      return spec.bind(stateId, stateAccessor);
+    }
+
+    @Override
+    public org.apache.beam.sdk.state.Timer timer(String timerId) {
+      checkState(
+          currentTimer.getValue() instanceof KV,
+          "Accessing timer in unkeyed context. Current timer is not a KV: %s.",
+          currentTimer);
+
+      return new FnApiTimer(timerId, (WindowedValue) currentTimer);
+    }
+
+    @Override
+    public PipelineOptions getPipelineOptions() {
+      return context.pipelineOptions;
+    }
+
+    @Override
+    public PipelineOptions pipelineOptions() {
+      return context.pipelineOptions;
+    }
+
+    @Override
+    public void output(OutputT output) {
+      outputTo(
+          mainOutputConsumers,
+          WindowedValue.of(output, currentTimer.getTimestamp(), currentWindow, PaneInfo.NO_FIRING));
+    }
+
+    @Override
+    public void outputWithTimestamp(OutputT output, Instant timestamp) {
+      checkArgument(
+          !currentTimer.getTimestamp().isAfter(timestamp),
+          "Output time %s can not be before timer timestamp %s.",
+          timestamp,
+          currentTimer.getTimestamp());
+      outputTo(
+          mainOutputConsumers,
+          WindowedValue.of(output, timestamp, currentWindow, PaneInfo.NO_FIRING));
+    }
+
+    @Override
+    public <T> void output(TupleTag<T> tag, T output) {
+      Collection<FnDataReceiver<WindowedValue<T>>> consumers =
```

Comment:
Should we be using a list rather than collection now that we've updated the multimap's interface? Ordering is not technically relevant here, so I'm not sure the change is appropriate anyway.

Reply:
Doesn't matter since its local to the class and not propagated elsewhere.